### ISSUE TITLE
Ambiguous class resolution false positive with Flexible Heredoc and Nowdoc Syntax

### ISSUE THREAD

**Issue by bcremer** [composer/composer#8080]
My `composer.json`:

```json
{
  "autoload": {
    "psr-4": {
      "Example\\": ""
    }
  }
}
```

File `DiffCommand.php`:

```php
<?php
namespace Example;
class DiffCommand
{
    const TEMPLATE = <<<'EOT'
        class Version<version> extends AbstractMigration
        EOT;
}
```

File `GenerateCommand.php`:
```php
<?php
namespace Example;
class GenerateCommand
{
    const TEMPLATE = <<<'EOT'
        class Version<version> extends AbstractMigration
        EOT;
}
```

Output of `composer diagnose`:

```
Checking composer.json: WARNING
No license specified, it is recommended to do so. For closed-source software you may use "proprietary" as license.
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 1.8.4
PHP version: 7.3.4
PHP binary path: /usr/bin/php
```

When I run this command:

```
composer dump-autoload --optimize
```

I get the following output:

```
Generating optimized autoload files

Warning: Ambiguous class resolution, "Example\Version" was found in both "/repro/DiffCommand.php" and "/repro/GenerateCommand.php", the first will be used.

Generated optimized autoload files containing 3 classes 
```

```
~/www/hotelapi/repro (HOAPI-13219-add-testbuidls-scripts) 
$ cat vendor/composer/autoload_classmap.php 
[...}
return array(
    'Example\\DiffCommand' => $baseDir . '/DiffCommand.php',
    'Example\\GenerateCommand' => $baseDir . '/GenerateCommand.php',
    'Example\\Version' => $baseDir . '/DiffCommand.php',
);
```

There should be no autoload entry for class `Example\\Version`.

**Comment by alcohol** [comment]
I guess this regex needs to be updated to take into account the changes that php 7.3 introduced:

https://github.com/composer/composer/blob/88852fbf1adfc1ffeb91e2db7cbbe1aa7f0321f9/src/Composer/Autoload/ClassMapGenerator.php#L164-L165

**Comment by Seldaek** [comment]
Fixed by #8082
